### PR TITLE
Bug 5383: handleNegotiationResult() level-2 debugs() crash

### PR DESCRIPTION
**Pull request by rousskov**
Writing a nil c-string to an std::ostream object results in undefined
behavior. When Security::IoResult::errorDescription is nil, that UB
leads to crashes for some STL implementations. These changes avoid UB
while making higher-level reporting code simpler and safer.

This change alters affected level-1 ERROR test lines a little, including
removing duplicated connection information from clientNegotiateSSL()
message (cache_log_message id=62). That duplication existed because
Squid reports the same Connection info automatically via CodeContext.

New WithExtras() mechanism may be useful for other "low-level debugging
and level-0/1 details for admins ought to differ" cases as well. Today,
the only known debugging context for Security::IoResult is
Security::PeerConnector::suspendNegotiation(), but that is likely to
change as we upgrade TLS callers to library-independent wrappers beyond
the current Security::Accept() and Security::Connect() pair.

### CODE REVIEW

File: src/client_side.cc
Comment (rousskov) [inline src/client_side.cc:2293]:
```diff
- ERROR: failure while accepting a TLS connection on conn41 local=127.0.0.1:21443 remote=127.0.0.1:55000 FD 25 flags=1: SQUID_TLS_ERR_ACCEPT+TLS_LIB_ERR=A00009C+TLS_IO_ERR=1
-     connection: conn41 local=127.0.0.1:21443 remote=127.0.0.1:55000 FD 25 flags=1
+ ERROR: Cannot accept a TLS connection
+     problem: failure
+     error detail: SQUID_TLS_ERR_ACCEPT+TLS_LIB_ERR=A00009C+TLS_IO_ERR=1
+     connection: conn40 local=127.0.0.1:21443 remote=127.0.0.1:54910 FD 20 flags=1
```

Client-to-Squid connection is still reported here (via CodeContext) even though it is not visible in new code. That is likely to (automatically) change as we start allocating or preserving master transaction context earlier, but current PR reporting is OK.

File: src/client_side.cc
Comment (rousskov) [inline src/client_side.cc:3041]:
```diff
- ERROR: failure while SslBump-accepting a TLS connection on conn35 local=127.0.0.1:3130 remote=127.0.0.1:53250 FD 20 flags=1: SQUID_TLS_ERR_ACCEPT+TLS_LIB_ERR=A00018A+TLS_IO_ERR=1
-     current master transaction: master53
+ ERROR: Cannot SslBump-accept a TLS connection
+     problem: failure
+     error detail: SQUID_TLS_ERR_ACCEPT+TLS_LIB_ERR=A00018A+TLS_IO_ERR=1
+     current master transaction: master54
```

No client-to-Squid connection reporting here because we do have a master transaction in this context. Admins are supposed to correlate that with their access.log entries if they need more information. We are using the same transaction/context-based approach for other level-0/1 messages as well.

File: src/security/PeerConnector.cc
Comment (rousskov) [inline src/security/PeerConnector.cc:284]: The above debugs() is where the reported bug was triggered, but all three debugs() have similar dangerous/buggy code.

```diff
- 83,2| PeerConnector.cc(281) handleNegotiationResult: ERROR: Cannot establish a TLS connection to conn38 local=127.0.0.1:35992 remote=127.0.0.1:443 HIER_DIRECT FD 22 flags=1:
-     problem: failure
-     detail: SQUID_TLS_ERR_CONNECT+TLS_LIB_ERR=A000410+TLS_IO_ERR=1
+ 83,2| PeerConnector.cc(281) handleNegotiationResult: ERROR: Cannot establish a TLS connection
+     problem: failure
+     error detail: SQUID_TLS_ERR_CONNECT+TLS_LIB_ERR=A000410+TLS_IO_ERR=1
+     connection: conn38 local=127.0.0.1:58424 remote=127.0.0.1:443 HIER_DIRECT FD 22 flags=1
```

We do want to report Squid-to-peer connection in such contexts because, unlike the client-to-Squid connection discussed elsewhere in this review, there may be multiple Squid-to-peer connections per master transaction and our access.log records do not record their details.

Please note that this is a level-2 message in current code. There is a corresponding TODO to upgrade "important" messages to level-1, and the message is already formatted as a level-0/1 message, so I used WithExtras() here.

**Review by rousskov — COMMENTED** [review]
The comments in this review show the difference between old and new debugs() messages based on a few sample errors. Three different error reporting contexts are related to this bug, although only the third one is known to be trigger crashes (in some cases).

File: src/security/PeerConnector.cc
Comment (yadij) [inline src/security/PeerConnector.cc:282]: So here we have the actual bug. This line should be displaying `result`, not just the `result.errorDescription` member.

None of the rest of this PR is relevant to fixing the bug, it is all just code bloat and obfuscation.

**Review by yadij — CHANGES_REQUESTED** [review]

File: src/security/PeerConnector.cc
Comment (rousskov) [reply src/security/PeerConnector.cc:282]: > So here we have the actual bug. 

Yes, for some definition of an "actual bug". IMO, there are actually several relevant bugs, and none of them are confined to this code line.

> This line should be displaying `result`, not just the `result.errorDescription` member.

This line should not display `result` as-is (in the official code) because doing so will produce difficult-to-grok high-level cache.log output -- existing Security::IoResult::print() is not meant to be used as-is in such contexts. Doing so will also make code duplication worse. This PR refactors Security::IoResult::print() to make simple result printing usable here (among other changes such as reducing code duplication).

> None of the rest of this PR is relevant to fixing the bug, it is all just code bloat and obfuscation.

This PR reduces the probability of a very similar "actual bug" from occurring: The other two debugs() statements have the same buggy code that will trigger the same crashes if seemingly unrelated code is added or changed. The existing buggy debugs() statements are also likely to be copied to other contexts as we upgrade TLS code, increasing the bug surface.

More consistent, safer, and simpler high-level code is _not_ "code bloat or obfuscation". This PR is an improvement even if there were no underlying "real bugs" to fix here! If you can achieve better results with less code, please do so (or suggest specific PR changes to arrive at that state instead). Otherwise, please remove your negative vote.

P.S. WithExtras() addition should be evaluated for "code bloat". I have already done that in this PR description (there are more details in branch commit 518d1307) and concluded that this addition is warranted. No flaws in that logic have been identified so far.

**Review by kinkie — APPROVED** [review]
While the solution is not minimal for this specific bug, it does make sense from a logical point of view that we would print the object and not just its description at different levels detail as decided by the calling context.

I considered what could an alternative look like, it could be for instance two debugs() statements with different levels inspecting the object or new getters, but this is brittle and lots of duplicated code.

Every other change is boilerplate to make it happen with standard c++ semantics; it is well contained and doesn't spill over, I don't see why this shouldn't go it.

@yadij, if you don't have any specific objections I'd be inclined to dismiss your request for changes and land this; please speak up before I move ahead

**Comment by rousskov** [comment]
> Or @rousskov could reduce this PR to the necessary minimal (bug 5383) fix appropriate for porting to all the older versions and packages that need it.

While it is possible to first avoid crashes in thee known problematic callers by duplicating code and then essentially undo that bad fix to implement a proper one (i.e. this PR), I should not be _required_ to do so.

This bug can be fixed in older Squid versions by writing duplicated code or by backporting this change. There are non-trivial tradeoffs involved in that choice; it is up to older version maintainers to decide how to handle that (and I am willing to help if needed). This master/v7 PR should be allowed to go in even if maintainers would prefer to just backport a problematic fix instead, especially after this PR has been waiting for 3+ months!

**Review by yadij — COMMENTED** [review]
Not sure why GH marks me a "requested changes". I explicitly did not do a review earlier, just feedback comments.

**Comment by rousskov** [comment]
> Not sure why GH marks me a "requested changes".

You explicitly [requested changes](https://github.com/squid-cache/squid/pull/1856#discussion_r1660892598) in response to my comment on July 1, 2024. After that, you left comments (without requesting changes), but that original negative vote naturally stays negative until it is explicitly dismissed or you approve PR. Thank you for dismissing it now.

FWIW, writing a top-level comment when requesting changes would make that change request to stand out more on the primary PR page (rather than be buried deeper inside an (often poorly correlated) comment thread).

**Pull request closed without merging**
